### PR TITLE
Mineral magnet tgui

### PR DESCRIPTION
**Pull request by codeJester27**
## About the PR 
Gives the mineral magnet a new tgui menu, accessed by the mineral magnet controls computer or the pod magnet link thingy.
imageimage
imageimage
image

## Why's this needed? 
tgui. Now it auto-updates, you can geo scan at any time instead of having to wait for the magnet to finish pulling, and there are probably less href exploits that nobody actually used anyways.

## Changelog

```changelog
(u)CodeJester
(*)Mineral magnets, how do they work? Maybe you can find out with its new UI menu.
```

### CODE REVIEW

File: code/obj/mining.dm [hunk 1]
```diff
@@ -701,161 +702,82 @@
...
+				miningEncounters += list(list(
+					name = encounter.name,
+					id = encounter_id
+				))
+		.["miningEncounters"] = miningEncounters
+
+		.["time"] = world.time
+
+	ui_act(action, params)
+		var/magnetNotReady = src.active || (src.last_used > world.time && !src.cooldown_override) || src.last_use_attempt > world.time
```
Comment (ZeWaka) [inline code/obj/mining.dm:726]: we should convert to TIME while we're here i think

**Comment by codeJester27** [comment]
> Neato! Does the telescope location list automatically expand the window? Otherwise you might want to make it a little longer since only three lines of scrolling is kinda cramped.

I just changed it to three and a half lines to make it more obvious that you can scroll down for more locations, still might make it taller if needed.

File: code/obj/mining.dm [hunk 2]
```diff
@@ -701,161 +702,82 @@
...
+
+	ui_act(action, params)
+		var/magnetNotReady = src.active || (src.last_used > TIME && !src.cooldown_override) || src.last_use_attempt > TIME
+		switch(action)
+			if ("geoscan")
+				var/MC = src.get_magnetic_center()
+				if (!MC)
+					boutput(usr, "Error. Magnet is not magnetized.")
+					return
+				mining_scan(MC, usr, src.get_scan_range())
```
Comment (mordent-goonstation) [inline code/obj/mining.dm:733]: Sanity check on spamming this button and what that would do. Should we have *some* kind of cooldown (even if it's just 1 second)?

File: code/obj/mining.dm [hunk 3]
```diff
@@ -701,161 +702,82 @@
...
-		else if (href_list["geo_scan"])
-			var/MC = src.get_magnetic_center()
-			if (!MC)
-				boutput(usr, "Error. Magnet is not magnetized.")
-				src.updateUsrDialog()
-				return
+					src.last_use_attempt = TIME + 10 // This is to prevent href exploits or autoclickers from pulling multiple times simultaneously
+					SPAWN(0)
+						if (src)
+							src.pull_new_source()
```
Comment (mordent-goonstation) [inline code/obj/mining.dm:762]: Sanity check on why we need to bump this out?

File: tgui/packages/tgui/interfaces/MineralMagnet.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+          </Box>
+        </Section>
+        {isLinked ? (
+          <Section title="Magnet Controls">
+            <Button
+              textAlign="center"
+              color="teal"
+              icon="rss"
+              onClick={() => act('geoscan')}
+              fluid >
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/MineralMagnet.js:59]:
```suggestion
              fluid>
```

File: tgui/packages/tgui/interfaces/MineralMagnet.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                <Dimmer fontSize={1.75} pb={2}>
+                  {magnetActive ? "Magnet Active" : "On Cooldown"}
+                </Dimmer>
+              )}
+              <Button
+                textAlign="center"
+                color={onCooldown && magnetCooldownOverride ? "average" : "purple"}
+                icon="magnet"
+                onClick={() => act('activatemagnet')}
+                fluid >
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/MineralMagnet.js:73]:
```suggestion
                fluid>
```

File: tgui/packages/tgui/interfaces/MineralMagnet.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                      onClick={() => act('activateselectable', { encounter_id: encounter.id })}
+                      fluid>
+                      {encounter.name}
+                    </Button>
+                  ))}
+                </Section>
+              </Collapsible>
+              <Button.Checkbox checked={magnetCooldownOverride}
+                onClick={() => act('overridecooldown')}
+                style={{ 'z-index': '2' }}>
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/MineralMagnet.js:91]: Why is z-index needed? I suspect it's not (as I believe it's added in React via the `style` prop with the key `zIndex`).

File: tgui/packages/tgui/interfaces/MineralMagnet.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                Override Cooldown
+              </Button.Checkbox>
+              <Button.Checkbox checked={magnetAutomaticMode}
+                onClick={() => act('automode')}
+                style={{ 'z-index': '2' }}>
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/MineralMagnet.js:96]: Why is z-index needed? I suspect it's not (as I believe it's added in React via the `style` prop with the key `zIndex`).

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]
Overall opinion: the UI has a lot of seemingly random colors being set for buttons (that also change?)

The visual style here is way off because of that, and I'd like to remove at least most of those (certainly the louder ones like teal/purple).

This is a first pass only, I'm sure I'll have more.

File: code/obj/mining.dm (hunk 2 above)
Comment (codeJester27) [reply code/obj/mining.dm:733]: The handheld geological scanner doesn't have a cooldown, and this button can't mess up the magnet when spammed like "Activate Magnet/telescope location" can, so not really sure.

File: tgui/packages/tgui/interfaces/MineralMagnet.js (hunk 6 above)
Comment (codeJester27) [reply tgui/packages/tgui/interfaces/MineralMagnet.js]: I really don't know why but I can't get `zIndex` to work in place of `z-index`. Maybe I'm doing it wrong (in which case I would appreciate a code suggestion) but only `z-index` actually does what I need here, which is making the checkboxes appear over the "Magnet Active" or "On Cooldown" dimmers.
image

**Comment by codeJester27** [comment]
Changed up the design.

File: code/obj/mining.dm (hunk 3 above)
Comment (codeJester27) [reply code/obj/mining.dm:762]: I need a bit more clarification

**Comment by ZeWaka** [comment]
!merge_upstream

File: code/obj/mining.dm (hunk 3 above)
Comment (ZeWaka) [reply code/obj/mining.dm:762]: why spawn

**Comment by ZeWaka** [comment]
!merge_upstream

File: browserassets/tgui/tgui.bundle.js [hunk 8]
```diff
@@ -1 +1 @@
...),(0,o.createComponentVNode)(2,c.Stack.Item,{width:"50%",children:(0,o.createComponentVNode)(2,c.Button,{icon:"skull-crossbones",fluid:!0,selected:m,onClick:function(){return N(!0)},children:"Lethal"})})]})})]}),!p&&!!C&&(0,o.createComponentVNode)(2,c.Section,{children:"Panel locked, swipe ID card to unlock."}),!!p&&(0,o.createComponentVNode)(2,c.Box,{py:"20px",children:[(0,o.createComponentVNode)(2,c.Box,{align:"center",fontFamily:"Courier New",children:(0,l.glitch)("ERROR: UNABLE TO READ AUTHORIZATION",12)}),(0,o.createComponentVNode)(2,c.Box,{align:"center",style:{"font-size":"20px"},children:d(7)})]})]})})})}},2861:function(e,t,n){"use strict";t.__esModule=!0,t.WeaponVendor=void 0;var o=n(39812),r=n(2497),a=n(71494),c=n(74814),i=n(85952),l=n(67113);t.WeaponVendor=function(e,t){var n=(0,a.useBackend)(t).data,r=(0,a.useLocalState)(t,"filter-available",!1),u=r[0],s=r[1];return(0,o.createComponentVNode)(2,i.Window,{width:550,height:700,children:(0,o.createComponentVNode)(2,i.Window.Content,{children:(0,o.createComponentVNode)(2,c.Stack,{className:"WeaponVendor",vertical:!0,fill:!0,children:[(0,o.createComponentVNode)(2,c.Stack.Item,{children:(0,o.createComponentVNode)(2,c.Section,{
```
Comment (ZeWaka) [inline browserassets/tgui/tgui.bundle.js:1]:
```suggestion
!function();
```
my bad on the resolve

**Comment by ZeWaka** [comment]
!merge_upstream

File: code/obj/mining.dm (hunk 2 above)
Comment (mordent-goonstation) [reply code/obj/mining.dm:733]: Handheld one also only affects one tile, I believe? Been a while since I've used it.

**Comment by ZeWaka** [comment]
!merge_upstream

File: code/obj/mining.dm (hunk 2 above)
Comment (codeJester27) [reply code/obj/mining.dm:733]: You can use the handheld one to scan a single tile for more details, but using it in-hand does a scan with the same range as the mining magnet.

**Comment by ZeWaka** [comment]
bad bot bad

**Comment by ZeWaka** [comment]
!merge_upstream

File: tgui/packages/tgui/interfaces/MineralMagnet.js (hunk 6 above)
Comment (mordent-goonstation) [reply tgui/packages/tgui/interfaces/MineralMagnet.js]: Honestly unsure about this, maybe a difference in Inferno vs. React or something. If your implementation works then I'm fine with using `z-index` over `zIndex`.

File: code/obj/mining.dm (hunk 2 above)
Comment (mordent-goonstation) [reply code/obj/mining.dm:733]: Fair enough. I view this as a slight problem, but it's not one you're introducing here and can be resolved independently.

**Review by mordent-goonstation — APPROVED** [review]
Looks like will need manually merging with upstream as this is before the workflow `stale.yml` change or something? If that gets resolved then consider good to merge.

**Pull request merged**